test/integration/bastion/bastion_test.go

```diff
@@ -0,0 +1,609 @@
+
+	err := routers.Delete(openstackClient.NetworkingClient, routerID).ExtractErr()
+	Expect(err).NotTo(HaveOccurred())
+
+	logger.Infof("Router '%s' is deleted...", routerID)
+	return nil
+}
+
+func prepareNewNetwork(logger *logrus.Entry, networkName string, openstackClient *OpenstackClient) (*string, error) {
+	logger.Infof("Waiting until network '%s' is created...", networkName)
+
+	network, err := networks.Create(openstackClient.NetworkingClient, networks.CreateOpts{Name: networkName}).Extract()
+	Expect(err).NotTo(HaveOccurred())
+
+	logger.Infof("Network '%s' is created...", networkName)
+	return &network.ID, nil
+}
+
+func prepareSubNet(logger *logrus.Entry, subnetName, networkid string, openstackClient *OpenstackClient) (*string, error) {
+	logger.Infof("Waiting until Subnet '%s' is created...", subnetName)
+
+	createOpts := subnets.CreateOpts{
+		Name:      subnetName,
+		NetworkID: networkid,
+		IPVersion: 4,
+		CIDR:      "10.180.0.0/16",
+		GatewayIP: pointer.String("10.180.0.1"),
+		AllocationPools: []subnets.AllocationPool{
+			{
+				Start: "10.180.0.2",
+				End:   "10.180.255.254",
+			},
+		},
+	}
+	subnet, err := subnets.Create(openstackClient.NetworkingClient, createOpts).Extract()
+	Expect(err).NotTo(HaveOccurred())
+	logger.Infof("Subnet '%s' is created...", subnetName)
+	return &subnet.ID, nil
+}
+
+// prepareShootSecurityGroup create fake shoot security group which will be used in EgressAllowSSHToWorker remoteGroupID
+func prepareShootSecurityGroup(logger *logrus.Entry, shootSgName string, openstackClient *OpenstackClient) (*string, error) {
+	logger.Infof("Waiting until Shoot Security Group '%s' is created...", shootSgName)
+
+	sgroups, err := groups.Create(openstackClient.NetworkingClient, groups.CreateOpts{Name: shootSgName, Description: shootSgName}).Extract()
+	Expect(err).NotTo(HaveOccurred())
+	logger.Infof("Shoot Security Group '%s' is created...", shootSgName)
+	return &sgroups.ID, nil
+}
+
+func teardownShootSecurityGroup(logger *logrus.Entry, groupID string, openstackClient *OpenstackClient) error {
+	err := groups.Delete(openstackClient.NetworkingClient, groupID).ExtractErr()
+	Expect(err).NotTo(HaveOccurred())
+	logger.Infof("Shoot Security Group '%s' is deleted...", groupID)
+	return nil
+}
+
+func teardownNetwork(logger *logrus.Entry, networkID string, openstackClient *OpenstackClient) error {
+	logger.Infof("Waiting until network '%s' is deleted...", networkID)
+
+	allPages, err := ports.List(openstackClient.NetworkingClient, ports.ListOpts{NetworkID: networkID}).AllPages()
+	Expect(err).NotTo(HaveOccurred())
+
+	allPorts, err := ports.ExtractPorts(allPages)
+	Expect(err).NotTo(HaveOccurred())
+
+	for _, port := range allPorts {
+		_, err = ports.Update(openstackClient.NetworkingClient, port.ID, ports.UpdateOpts{DeviceOwner: pointer.String("")}).Extract()
+		Expect(err).NotTo(HaveOccurred())
+		err = ports.Delete(openstackClient.NetworkingClient, port.ID).ExtractErr()
+		Expect(err).NotTo(HaveOccurred())
+	}
```

Comment:
This could be replaced with 

```
func teardownNetwork(logger *logrus.Entry, networkID, routerID, subnetID string, openstackClient *OpenstackClient) error {
	logger.Infof("Waiting until network '%s' is deleted...", networkID)

	_, err := routers.RemoveInterface(openstackClient.NetworkingClient, routerID, routers.RemoveInterfaceOpts{
		SubnetID: subnetID,
	}).Extract()
	Expect(err).NotTo(HaveOccurred())
```
Could you give a try like that ?

Reply:
after the test seems not to work. the private network will be left in the Networks & Routers with subnet and ports resource.
the `RemoveInterface` method seems only to detach the subnet from the router, to remove the subnet still need to get out of ports first otherwise it displays ports still in use

Reply:
The `RemoveInterface` should remove the router port one of the ports that are not cleaned so far by the cleanup. We used the removing of the device_owner in the past as a workround for old LB ports that couldn't be deleted but this is not an old LB port, so there are things missing in the cleanup logic. What other ports are still left ?

Reply:
two DHCP ports `Attached Device: network:dhcp` which are generated automatically when subnet/ private network is created.

Reply:
some misconfiguration on my local code. help from @kon-angelo  it works now , Thanks!
3d91a5dd